_posts/2023-06-27-redis.md

```diff
@@ -0,0 +1,23 @@
+
+Another major concern with redis is durability. Redis out of the box does not persist data on disk, only in memory. This means, when a server goes down, so does all you data. Durability is serious business and when it becomes a priority. And this is where redis starts to go backwards. Redis was never planned to be provide durability beyond RAM. This is evident in the fact that disk persistence, was never part of redis until, v 0.04. Redis supports two types of [persistence modes](https://redis.io/docs/management/persistence/).
+
+The first one is called snapshotting or RDB. When snapshotting is enabled redis will periodically write all your dataset in memory to disk. This is good for point in time recovery. But this also means, you loose all the data between when the last snapshot was created versus when the failure occurred. For a moderately busy server there are bound to be significant changes that happen between when snapshots are set, and loosing it might not be a good thing. To combat this, some teams, set the snapshots times as minimal as possible to combat the amount of data it is possible to loose. This can be a bad idea, when your data set is considerably large. Writing a 1gig file to disk every every 60 seconds, is a recipe for disaster. [What happens in the background](https://redis.io/docs/getting-started/faq/#background-saving-fails-with-a-fork-error-on-linux) is redis forks a child process for background processing, serializing the dataset in memory, making it disk compatible, writing it to a temporary file in the background and rename the file atomically upon finish. Even though the overhead of creating a fork is zero in theory when the OS supports copy-on-write, you still need to turn on the `overcommit_memory`. This is because if the dataset between the parent process and the child process deviate, redis will not keep track of the changes and will have to allocate just as much memory your data set has to the child process in order for the snapshot to be successful. In snapshotting mode, you must do everything to ensure that, your datastore does not exceed half the RAM allocated to redis, otherwise your redis server will implode with an OOM. This is the default persistence mode cos it is simple and safe for small data sets. When you data set starts to increase considerably in size, think twice about snapshots. 
```

Comment:
```suggestion
The first one is called snapshotting or RDB. When snapshotting is enabled, Redis will periodically write all your dataset in memory to disk. This is good for point-in-time recovery. But this also means you lose all the data between when the last snapshot was created versus when the failure occurred. For a moderately busy server, there are bound to be significant changes that happen between when snapshots are set, and losing it might not be a good thing. To combat this, some teams set the snapshot times as minimal as possible to combat the amount of data it is possible to lose. This can be a bad idea when your data set is considerably large. Writing a 1gig file to disk every 60 seconds is a recipe for disaster. [What happens in the background](https://redis.io/docs/getting-started/faq/#background-saving-fails-with-a-fork-error-on-linux) is Redis forks a child process for background processing, serializing the dataset in memory, making it disk compatible, writing it to a temporary file in the background and rename the file atomically upon finish. Even though the overhead of creating a fork is zero in theory when the OS supports copy-on-write, you still need to turn on the `overcommit_memory`. This is because if the dataset between the parent process and the child process deviates, Redis will not keep track of the changes and will have to allocate just as much memory your data set has to the child process in order for the snapshot to be successful. In snapshotting mode, you must do everything to ensure that your datastore does not exceed half the RAM allocated to Redis, otherwise, your Redis server will implode with an OOM. This is the default persistence mode because it is simple and safe for small data sets. When your data set starts to increase considerably in size, think twice about snapshots. 
```